system/jlib/jptree.cpp

```diff
@@ -3176,6 +2818,408 @@ bool PTree::checkPattern(const char *&xxpath) const
+        }
+    }
+    else
+    {
+        while (a-- != attrs)
+        {
+            if (streq(a->key->get(), key))
+                return a;
+        }
+    }
+    return nullptr;
+}
+
+const char *CAtomPTree::getAttributeValue(const char *key) const
+{
+    AttrValue *e = (AttrValue *)findAttribute(key);
+    if (e)
+        return e->value->get();
+    return nullptr;
+}
+
+unsigned CAtomPTree::getAttributeCount() const
+{
+    return numAttrs;
+}
+
+const void *CAtomPTree::getNextAttribute(const void *_cur, AttrKeyValue &kv) const
+{
+    if (nullptr == _cur)
+    {
+        if (attrs)
+        {
+            kv.key = attrs->key->get();
+            kv.value = attrs->value->get();
+        }
+        return attrs;
+    }
+    else
+    {
+        AttrValue *end = attrs+numAttrs;
+        AttrValue *cur = (AttrValue *)_cur;
+        if (cur+1 == end)
+            return nullptr;
+        ++cur;
+        kv.key = cur->key->get();
+        kv.value = cur->value->get();
+        return cur;
+    }
+}
+
+void CAtomPTree::setAttribute(const char *key, const char *val)
+{
+    if (!key)
+        return;
+    if (!validateXMLTag(key+1))
+        throw MakeIPTException(-1, "Invalid xml attribute: %s", key);
+    if (!val)
+        val = "";  // cannot have NULL value
+    AttrValue *a = attrs+numAttrs;
+    AttrValue *v = nullptr;
+    bool nc = isnocase();
+    if (nc)
+    {
+        while (a-- != attrs)
+        {
+            if (strieq(a->key->get(), key))
+            {
+                v = a;
+                break;
+            }
+        }
+    }
+    else
+    {
+        while (a-- != attrs)
+        {
+            if (streq(a->key->get(), key))
+            {
+                v = a;
+                break;
+            }
+        }
+    }
+    if (v)
+    {
+        if (streq(v->value->get(), val))
+            return;
+        CriticalBlock block(hashcrit);
+        attrHT->removeval(v->value);
+        v->value = attrHT->addval(val);
+    }
+    else
+    {
+        CriticalBlock block(hashcrit);
+        AttrValue *newattrs = newAttrArray(numAttrs+1);
+        memcpy(newattrs,attrs, numAttrs*sizeof(AttrValue));
+        newattrs[numAttrs].key = attrHT->addkey(key, nc);
+        newattrs[numAttrs].value = attrHT->addval(val);
+        freeAttrArray(attrs, numAttrs);
+        numAttrs++;
+        attrs = newattrs;
+    }
+}
+
+bool CAtomPTree::removeAttribute(const char *key)
+{
+    AttrValue *a = attrs+numAttrs;
+    AttrValue *del = nullptr;
+    if (isnocase())
+    {
+        while (a-- != attrs)
+        {
+            if (strieq(a->key->str, key))
+            {
+                del = a;
+                break;
+            }
+        }
+    }
+    else
+    {
+        while (a-- != attrs)
+        {
+            if (streq(a->key->str, key))
+            {
+                del = a;
+                break;
+            }
+        }
+    }
+    if (!del)
+        return false;
+    numAttrs--;
+    CriticalBlock block(hashcrit);
```

Comment:
This is a shame... removing an attribute needs to lock and clone all the attributes.